keps/sig-auth/20190607-certificates-api.md

```diff
@@ -132,6 +133,75 @@ A typical successful issuance proceeds as follows.
    the `.Status.Certificate` field.
 1. The requestor observes the update, and stores the certificate locally.
 
+### Signers
+CSRs have a `signerName` field which is used to specify which signer the CSR creator wants to sign the certificate.
+To support migration from v1beta1 to v1, this required field will be defaulted in v1beta1 (optional in openapi), but
+not defaulted and required in v1 :
+ 1. If it's a kubelet client certificate, it is assigned "kubernetes.io/kubelet-client".
+ 2. Otherwise, it is assigned "kubernetes.io/legacy-unknown".
+
+There will be field selector support to make approvers and signers easier to write.
+
+All signers should provide information about how they work so that clients can predict what will happen to their CSRs.
```

Comment:
In my experience working with different types of CAs, some of this information is not as easy to 'come by' as I'd like. We often find CAs doing things like:

* Automatically promoting a `dnsNames[0]` to be a `commonName`
* Using different CAs for different certificates (although this is usually due to some kind of 'profile' or 'zone' attribute being set on the request for a certificate)
* Defaulting `keyUsages`
* Ignoring `notBefore` and `notAfter` parameters

A lot of CA APIs additionally don't provide a clear way to determine which of these things they are going to do, so whilst I _am_ in favour of the `signers` being able to specify this, I think we *also* need to support *not* specifying this.

> so that clients can predict what will happen to their CSRs.

This sounds to me like we're trying to touch on policy control for CSRs (i.e. "all domains requested here must have a suffix of `example.com`) - this is definitely a valuable/useful thing, however I'm uncertain whether this is the best place for it to exist. We've though about `CertificatePolicy` in cert-manager in the past, and the general thinking is to separate out the policy from the CSR API and instead make it a property of the *approver*. This can be seen as 'shifting left' the policy decisions that a CA may take. This makes sense to me because the CA is going to reject a request *anyway* if it violates policy, and the signer plugin should update the CSR accordingly to make that clear.

Is the motivation for exposing this as a property of the signer so that a generic approval/policy enforcement plugin can be written? If so, I'd worry that we'll not expose enough flexibility to do this effectively (i.e. because in reality, one signer _may_ represent multiple different 'types' of certificate, and each of these types may have their own different types of restriction)

Reply:
In short: if this is in fact a policy engine, I think we should be cautious to not create CertificateSecurityPolicy (a la PodSecurityPolicy). Policy is a complex thing, and different orgs may have very different types of policy applied, some of which may even be evaluated or defined externally.

Reply:
> A lot of CA APIs additionally don't provide a clear way to determine which of these things they are going to do, so whilst I am in favour of being able to specify this in some way, I think we also need to support not specifying this.

Sure, a definer for a particular signerName is free to explicitly say "behavior here is unspecified". That obviously makes that signerName less reliable/portable, but that is a choice a particular signer can make.

We would tightly define the meaning and behavior of the `kubernetes.io/...` signer names.

> This sounds to me like we're trying to touch on policy control for CSRs (i.e. "all domains requested here must have a suffix of example.com)

This isn't intended to be a policy engine, just a well-specified API, documenting what a particular signerName means/allows, so clients know what to expect when they request CSRs for that signerName (e.g. should they rely on getting all the SANs they requested? only the SANs they requested? should they rely on being able to request/get specific key usages?)

> the general thinking is to separate out the policy from the CSR API and instead make it a property of the approver

Most CA's I've encountered have bounds on what they will sign. If you make this a property of the approver, you also need to define what happens when the approver approves something the CA will not sign, and then you're back to specifying what the CA will sign.

Additionally, we want to enable specific clusters to use their own signers for the standard Kubernetes certificate types (kubelet serving cert, kube-apiserver client cert, etc), but those replacement signers need to behave consistently with kube-controller-manager. We just need to be very clear about which aspects of the resulting certificate are considered fundamental to the signerName.

Reply:
>  when the approver approves something the CA will not sign, and then you're back to specifying what the CA will sign.

I agree on this, and having the information up front is definitely useful. But in a lot of other settings (i.e. large CA management tools) it is not feasible to mirror this information into the signer API, meaning we will absolutely still need a standardised way to communicate "this certificate cannot be signed due to X issue" - this has been touched on/requested here for our own type: https://github.com/jetstack/cert-manager/issues/2289

How can a signer communicate to the consumer of the API that this is the reason the issuance failed?

Reply:
> we will absolutely still need a standardized way to communicate "this certificate cannot be signed due to X issue"
> ...
> How can a signer communicate to the consumer of the API that this is the reason the issuance failed?

That's a good point. Either a condition or a status field peer to the certificate field would make sense to me. I think defining how signers should report failure in a way that a client knows the signing request failed and understands whether they should retry or not should be done before v1, but is orthogonal to this multi-signer support.

Reply:
Sure - it's something we are about to explore how to represent in cert-manager soon, so hopefully there'll be some more insight to help us put another KEP together.